Feature/friendship v2

ClaytonSiby opened the pull request:
In this branch, we implemented the following:

- mutual friendships between users
- RSpec feature tests to send friend requests.

ClaytonSiby committed:
Add mutual friendships feature

ClaytonSiby committed:
Add feature tests for sending friend requests

ClaytonSiby committed:
Run rubocop tests

moinkhanif reviewed (changes requested):
## Changes Required ❗

Hey team, nice work on the project so far, a few changes are still required for the project to be considered.

- Friendship should only be created once. On the users index page, we are able to create multiple friendships for the same person.

![image](https://user-images.githubusercontent.com/6555800/97985296-1a980080-1dfe-11eb-8ce9-4eea88ba0a06.png)

- After friendship confirmation, the user is unable to view the friend's profile. Kindly fix that.

- `User#friend_requests` can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```
- Similar to `friend_requests` create `User#pending_friends` SQL query data instead of the ruby method.
- Also create a `friends` query instead of a ruby method.

After committing the required changes, please request for a review again.

**Happy coding**

ClaytonSiby committed:
Update all users page to allow user to send only one request to another user

ClaytonSiby committed:
Refactor ruby code to SQL queries in User model

ClaytonSiby commented:
We updated the views so that the app allows a user to send one friend request to each of the other users. We also refactored the ruby code to use SQL for database queries as requested by the REVIEW.

maelfosso reviewed (changes requested):
# Changes required 🔴 

Hello,

Thank you for implementing the suggested changes
However, there are some missing requirements and issues

1- There are some errors when running `bundle exec rspec`
```rb
Failures:

  1) CreateFriendships User creates a new friendship
     Failure/Error: Unable to find matching line in /home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/app/assets/stylesheets/application.scss
     
     ActionView::Template::Error:
       File to import not found or unreadable: bootstrap/scss/bootstrap.
       Load paths:
         /home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/app/assets/config
         /home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/app/assets/images
         /home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/app/assets/javascripts
         /home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/app/assets/stylesheets
         /home/maelfosso/.rvm/gems/ruby-2.7.0/gems/coffee-rails-4.2.2/lib/assets/javascripts
         /home/maelfosso/.rvm/gems/ruby-2.7.0/gems/actioncable-5.2.4.1/lib/assets/compiled
         /home/maelfosso/.rvm/gems/ruby-2.7.0/gems/activestorage-5.2.4.1/app/assets/javascripts
         /home/maelfosso/.rvm/gems/ruby-2.7.0/gems/actionview-5.2.4.1/lib/assets/compiled
         /home/maelfosso/.rvm/gems/ruby-2.7.0/gems/turbolinks-source-5.2.0/lib/assets/javascripts
         /home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/node_modules
  2) User associations is expected to have many inverse_friendships
     Failure/Error: it { should have_many(:inverse_friendships) }
       Expected User to have a has_many association called inverse_friendships (no association called inverse_friendships)
     # ./spec/models/user_spec.rb:14:in `block (3 levels) in <main>'

Finished in 6.69 seconds (files took 6.1 seconds to load)
10 examples, 2 failures
```

2- I got this error when launching the app
![Screenshot at 2020-11-03 17:35:54](https://user-images.githubusercontent.com/8848104/98013814-233b0780-1dfb-11eb-8cc7-3b5b709aea08.png)

3- Kindly remove logic from views (ex `views/users/show.html.erb` and `views/users/_user.html.erb`). Move them into view helpers

Kindly fix them and submit a new code review request
Thanks

Mael FOSSO

ClaytonSiby committed:
Update test example for user associations

ClaytonSiby commented:
Hi @maelfosso thank you for the review, we updated the test example which was failing the tests & please note, if you're running the app for the first time, you need to also run: `yarn install` or `yarn add bootstrap jquery popper.js` to add bootstrap dependencies. That is why you're having the `Sass::SyntaxError in Devise::Session#new`.

lekegitrepo reviewed (changes requested):
## Status: Request changes :recycle:
#### Good job so far :+1: However, some of the required changes are not well implemented. Importantly, the configuration of your database does not allow migration it is complaining about user name and password. Check below for more comments.

**_This explanation "`if you're running the app for the first time, you need to also run: yarn install or yarn add bootstrap jquery popper.js to add bootstrap dependencies`" that you added in the comment should be added to the readme instruction_**

#### _Reviewed by Temitayo Adeleke_

ClaytonSiby committed:
Update README.md file

ClaytonSiby committed:
Update README.md

ClaytonSiby committed:
Update README.md

Rumbie-Mudzie7 commented:
Thank you @lekegitrepo  for the timely and constructive review.  The necessary changes have been implemented as requested and please find in the updated Readme file(#Setup), the database configuration instructions.  This will solve the password and username prompt issues.

maelfosso reviewed (changes requested):
# Changes required 🔴 

Hello,

It's working now like a charm. Thank you 👍🏾 
However, there are still some issues

1- Kindly remove logic from views (ex views/users/show.html.erb and views/users/_user.html.erb). Move them into view helpers

I talking about this `if` statement
https://github.com/ClaytonSiby/ror-social-scaffold/blob/7822a563285724b8f1ba4c7b6d29578fa19653cb/app/views/users/show.html.erb#L3

2-  The `rubocop.yml` file in your project is different than the required one.
Kindly follow this [link](https://github.com/microverseinc/linters-config/tree/master/ror) to set up the good one, please

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

ClaytonSiby committed:
Update .rubocop.yml file to meet the requirements

ClaytonSiby committed:
Remove logic from the views

maelfosso commented on app/views/users/_non_friend.html.erb:
```diff
@@ -0,0 +1,39 @@
+<section class="container">
+  <% if non_friend?(@user) %>
+    <div class="card">
+      <div class="card-body">
+        <p class="card-title ">Name: <span class="text-warning user_name"><%= @user.name %></span></p>
+        <div class="">
+          <h5 class="recent_posts">Recent Posts: </h5>
+          <% @posts.each do |post| %>
+            <div class="card mb-4">
+              <div class="card-body">
+                  <p class="text-italic"><%= post.content %></p>
+              </div>
+              <div class="card-footer">
+                <p>Creation Date: <span class="italic"><%= time_ago_in_words(post.created_at) %></span></p>
+                <p>Post Info: <%= link_to 'View Post', root_path , class: 'btn btn-sm btn-success' %> </p>
+              </div>
+            </div>
+          <% end %>
+        </div>
+
+        <div class="mutual-friends">
+          <h4 class="text-italic">Mutual Friends: </h4>
+          <% @mutual_friends.each do |friend|%>
+            <p><%= friend.name %></p>
+          <% end %>
+        </div>
+      </div>
+      
+      <div class="card-footer">
+        <button class="btn btn-sm btn-success">
+          <%= link_to 'Invite to Friendship', send_request_path(@user.id), class: 'text-light' %>
+        </button>
+        <button class="btn btn-sm btn-dark">
+          <%= link_to 'Back', users_path, class: 'text-light'%>
+        </button>
+      </div>
+    </div>
+  <% end %>
```
Kindly remove that the logic. Move that to the view helpers.

Rumbie-Mudzie7 replied:
Thank you @maelfosso for the timely go ahead and the note of the changes required before merging.

maelfosso reviewed (approved):
# Congratulations 👏🏾 

Thank you for implementing the suggested changes.

You can merge it however before do it, kindly read the comments left below and implement them before merging. **REMOVE ALL THE LOGIC IN YOUR VIEW AND IMPLEMENT THEM IN VIEW HELPERS**

Thanks and Good continuation 🤝 

Mael FOSSO

ClaytonSiby committed:
Update single user page & remove logic from the views

ClaytonSiby committed:
Update users_helper method to determine friendship
